extensions/skyportal/skyportal/handlers/api/alert.py

```diff
@@ -490,28 +477,66 @@ def post(self, objectId):
-                "magsys": df.magsys.tolist(),
-                "filter": df.ztf_filter.tolist(),
-                "ra": df.ra.tolist(),
-                "dec": df.dec.tolist(),
-            }
+            # filter out bad data:
+            mask_good_diffmaglim = df["diffmaglim"] > 0
+            df = df.loc[mask_good_diffmaglim]
 
-            photometry_handler = PhotometryHandler(request=self.request, application=self.application)
-            photometry_handler.request.body = tornado.escape.json_encode(photometry)
-            try:
-                photometry_handler.put()
-            except Exception:
-                log(f"Failed to post photometry of {objectId}")
-            # do not return anything yet
-            self.clear()
+            # get group stream access and map it to ZTF alert program ids
+            group_stream_access = []
+            for group in groups:
+                group_streams = (
+                    DBSession()
+                    .query(Stream)
+                    .join(GroupStream)
+                    .filter(GroupStream.group_id == group.id)
+                    .all()
+                )
+                if group_streams is None:
+                    group_streams = []
+
+                group_stream_selector = {1}
+
+                for stream in group_streams:
+                    if "ztf" in stream.name.lower():
+                        group_stream_selector.update(set(stream.altdata.get("selector", [])))
+
+                group_stream_access.append(
+                    {
+                        "group_id": group.id,
+                        "permissions": list(group_stream_selector)
+                    }
+                )
+
+            # post data from different program_id's
+            for pid in set(df.programid.unique()):
+                group_ids = [gsa.get("group_id") for gsa in group_stream_access if pid in gsa.get("permissions", [1])]
+
+                if len(group_ids) > 0:
+                    pid_mask = df.programid == int(pid)
+
+                    photometry = {
+                        "obj_id": objectId,
+                        "group_ids": group_ids,
+                        "instrument_id": 1,  # placeholder
+                        "mjd": df.loc[pid_mask, "mjd"].tolist(),
+                        "mag": df.loc[pid_mask, "magpsf"].tolist(),
+                        "magerr": df.loc[pid_mask, "sigmapsf"].tolist(),
+                        "limiting_mag": df.loc[pid_mask, "diffmaglim"].tolist(),
+                        "magsys": df.loc[pid_mask, "magsys"].tolist(),
+                        "filter": df.loc[pid_mask, "ztf_filter"].tolist(),
+                        "ra": df.loc[pid_mask, "ra"].tolist(),
+                        "dec": df.loc[pid_mask, "dec"].tolist(),
+                    }
+
+                    if len(photometry.get('mag', ())) > 0:
+                        photometry_handler = PhotometryHandler(request=self.request, application=self.application)
```

Comment:
@dannygoldstein You've refactored similar code recently; do we have a better way than this of calling sub-handlers?  I thought we settled on a sign-in-via-generic-token approach.

Reply:
Sorry, my origin was outdated, fixed that. As for the approach, yes, we settled on it, but I haven't yet implemented it :)